commit 57433e4b74d6e7935a3585912e47982c494ca08e
Author: nephros <nemo@pgxperiiia10>
Date:   Tue Apr 28 09:13:16 2026 +0200

    [Sailfish OS] Report device as phone or tablet

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 88728957d..f5de3e771 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1793,6 +1793,13 @@ elseif(UNIX AND NOT APPLE AND NOT RISCOS AND NOT HAIKU)
     endif()
   endif()
 
+  if(SAILFISHOS)
+    sdl_glob_sources(
+      "${SDL3_SOURCE_DIR}/src/core/sailfishos/*.c"
+      "${SDL3_SOURCE_DIR}/src/core/sailfishos/*.h"
+    )
+  endif()
+
   if(SDL_AUDIO)
     if(NETBSD)
         set(SDL_AUDIO_DRIVER_NETBSD 1)
diff --git a/src/SDL.c b/src/SDL.c
index b1a1a5635..558c736f3 100644
--- a/src/SDL.c
+++ b/src/SDL.c
@@ -828,6 +828,18 @@ const char *SDL_GetPlatform(void)
 #endif
 }
 
+bool SDL_IsPhone(void)
+{
+#if defined(SDL_PLATFORM_ANDROID) || \
+    defined(SDL_PLATFORM_SAILFISHOS) || \
+    (defined(SDL_PLATFORM_IOS) && !defined(SDL_PLATFORM_VISIONOS))
+    if (!SDL_IsTablet() && !SDL_IsTV()) {
+        return true;
+    }
+#endif
+    return false;
+}
+
 bool SDL_IsTablet(void)
 {
 #ifdef SDL_PLATFORM_ANDROID
@@ -835,6 +847,8 @@ bool SDL_IsTablet(void)
 #elif defined(SDL_PLATFORM_IOS)
     extern bool SDL_IsIPad(void);
     return SDL_IsIPad();
+#elif defined(SDL_PLATFORM_SAILFISHOS)
+    return SDL_IsSailfishOSTablet();
 #else
     return false;
 #endif
diff --git a/src/core/sailfishos/SDL_sailfishos.c b/src/core/sailfishos/SDL_sailfishos.c
new file mode 100644
index 000000000..23a433ca4
--- /dev/null
+++ b/src/core/sailfishos/SDL_sailfishos.c
@@ -0,0 +1,60 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2026 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+
+#ifdef SDL_PLATFORM_SAILFISHOS
+
+#include "SDL_sailfishos.h"
+
+// new location starting from SFOS 5.1
+#define HW_RELEASE_FILENAME "/usr/lib/hw-release";
+#define HW_RELEASE_FILENAME_LEGACY "/etc/hw-release";
+
+bool SDL_IsSailfishOSTablet(void)
+{
+    bool ret = false;
+
+    FILE *file = fopen(HW_RELEASE_FILENAME, "r");
+    if (file == NULL) {
+        *file = fopen(HW_RELEASE_FILENAME_LEGACY, "r");
+        if (file == NULL) {
+            SDL_LogWarn(SDL_LOG_CATEGORY_SYSTEM, "Could not find hardware release file!";
+            return false;
+        }
+    }
+
+    char line[128];
+    while (fgets(line, sizeof(line), file) != NULL)
+    {
+        char dev[64];
+        SDL_sscanf(line, "MER_HA_DEVICE=%s", &dev);
+        if (( SDL_strcmp(dev, "tbj") == 0 )  // Original Jolla Tablet
+           || SDL_strcmp(dev, "pinetab")
+           || SDL_strcmp(dev, "pinetab2")
+           || SDL_strcmp(dev, "pipa") ) // VerandiTeams's port of Xiaomi Pad 6
+            { ret = true; break; }
+    }
+
+    return ret;
+}
+
+#endif // SDL_PLATFORM_SAILFISHOS
+
diff --git a/src/core/sailfishos/SDL_sailfishos.h b/src/core/sailfishos/SDL_sailfishos.h
new file mode 100644
index 000000000..a1a8e97d7
--- /dev/null
+++ b/src/core/sailfishos/SDL_sailfishos.h
@@ -0,0 +1,42 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2026 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+
+#ifndef SDL_sailfishos_h
+#define SDL_sailfishos_h
+
+// Set up for C function definitions, even when using C++
+#ifdef __cplusplus
+/* *INDENT-OFF* */
+extern "C" {
+/* *INDENT-ON* */
+#endif
+
+bool SDL_IsSailfishOSTablet(void);
+
+// Ends C function definitions when using C++
+#ifdef __cplusplus
+/* *INDENT-OFF* */
+}
+/* *INDENT-ON* */
+#endif
+
+#endif // SDL_sailfishos_h
