--- a/src/SDL.c	2026-03-03 21:10:43.000000000 +0100
+++ b/src/SDL.c	2026-04-14 13:18:30.988582280 +0200
@@ -835,6 +835,12 @@
 #elif defined(SDL_PLATFORM_IOS)
     extern bool SDL_IsIPad(void);
     return SDL_IsIPad();
+#elif defined(SDL_PLATFORM_SAILFISHOS)
+#if !SDL_VERSION_ATLEAST(3, 5, 0)
+        return true;
+#else
+#error "Please implement SDL_IsPhone() for SDL_PLATFORM_SAILFISHOS and update SDL_IsTablet()"
+#endif // SDL < 3.5.0
 #else
     return false;
 #endif
