commit 319712b9994e3650039fb81e7e33bdb7defc0ea2
Author: nephros <nemo@pgxperiiia10>
Date:   Tue Apr 28 09:14:13 2026 +0200

    [Sailfish OS] Detect SailJail as sandbox

diff --git a/include/SDL3/SDL_system.h b/include/SDL3/SDL_system.h
index ec23a1fe7..9b4f08607 100644
--- a/include/SDL3/SDL_system.h
+++ b/include/SDL3/SDL_system.h
@@ -653,7 +653,8 @@ typedef enum SDL_Sandbox
     SDL_SANDBOX_UNKNOWN_CONTAINER,
     SDL_SANDBOX_FLATPAK,
     SDL_SANDBOX_SNAP,
-    SDL_SANDBOX_MACOS
+    SDL_SANDBOX_MACOS,
+    SDL_SANDBOX_SAILJAIL,
 } SDL_Sandbox;
 
 /**
diff --git a/src/SDL.c b/src/SDL.c
index 3e61b8ad4..cd7861bba 100644
--- a/src/SDL.c
+++ b/src/SDL.c
@@ -868,6 +868,12 @@ static SDL_Sandbox SDL_DetectSandbox(void)
         return SDL_SANDBOX_SNAP;
     }
 
+#if defined(SDL_PLATFORM_SAILFISHOS)
+    if (access("/run/firejail/dbus/user", F_OK) == 0) {
+        return SDL_SANDBOX_SAILJAIL;
+    }
+#endif
+
     if (access("/run/host/container-manager", F_OK) == 0) {
         return SDL_SANDBOX_UNKNOWN_CONTAINER;
     }
