diff -Naur a/src/core/linux/SDL_progressbar.c b/src/core/linux/SDL_progressbar.c
--- a/src/core/linux/SDL_progressbar.c	2026-01-30 16:25:35.000000000 +0100
+++ b/src/core/linux/SDL_progressbar.c	2026-04-14 16:15:25.651248094 +0200
@@ -32,6 +32,11 @@
 #define UnityLauncherAPI_DBUS_INTERFACE "com.canonical.Unity.LauncherEntry"
 #define UnityLauncherAPI_DBUS_SIGNAL    "Update"
 
+// Lipstick is the window manager on Sailfish OS:
+#define LipstickLauncherAPI_DBUS_NODE "com.jolla.lipstick"
+#define LipstickLauncherAPI_DBUS_PATH "/LauncherModel"
+#define LipstickLauncherAPI_DBUS_INTERFACE "org.nemomobile.lipstick.LauncherModel"
+
 static char *GetDBUSObjectPath(void)
 {
     char *app_id = SDL_strdup(SDL_GetAppID());
@@ -106,6 +111,22 @@
         return false;
     }
 
+#ifdef SDL_PLATFORM_SAILFISHOS
+    char *desktop_path = GetAppDesktopPath();
+
+    if (!desktop_path) {
+        return false;
+    }
+    const int progress_visible = ShouldShowProgress(window->progress_state);
+    if(progress_visible) {
+        SDL_DBus_CallVoidMethod(LipstickLauncherAPI_DBUS_NODE, LipstickLauncherAPI_DBUS_PATH, LipstickLauncherAPI_DBUS_INTERFACE
+                                "notifyLaunching", DBUS_TYPE_STRING, desktop_path, DBUS_TYPE_INVALID);
+    } else {
+        SDL_DBus_CallVoidMethod(LipstickLauncherAPI_DBUS_NODE, LipstickLauncherAPI_DBUS_PATH, LipstickLauncherAPI_DBUS_INTERFACE
+                                "cancelNotifyLaunching", DBUS_TYPE_STRING, desktop_path, DBUS_TYPE_INVALID);
+    }
+    SDL_free(desktop_path);
+#else
     char *objectPath = GetDBUSObjectPath();
     if (!objectPath) {
         return false;
@@ -156,7 +177,7 @@
     SDL_free(desktop_path);
     dbus->message_unref(msg);
     SDL_free(objectPath);
-
+#endif
     return true;
 }
 
