commit dd0faf85ae9f8d4fe95d35076c1a6cc217b675f3
Author: nephros <nemo@pgxperiiia10>
Date:   Tue Apr 28 10:18:03 2026 +0200

    [Sailfish OS] Set pulseaudio media role

diff --git a/src/audio/pulseaudio/SDL_pulseaudio.c b/src/audio/pulseaudio/SDL_pulseaudio.c
index 2fcf078fe..1f708c24b 100644
--- a/src/audio/pulseaudio/SDL_pulseaudio.c
+++ b/src/audio/pulseaudio/SDL_pulseaudio.c
@@ -365,6 +365,18 @@ static bool ConnectToPulseServer(void)
         icon_name = "applications-games";
     }
     PULSEAUDIO_pa_proplist_sets(proplist, PA_PROP_APPLICATION_ICON_NAME, icon_name);
+#ifdef SDL_PLATFORM_SAILFISHOS
+    // re-use the hint from pipewire.
+    // if unset, use a Sailfish OS specific magic value, "x-maemo".
+    // if the role is not that, the OS will output audio at full volume,
+    // and offer no way to control it via mixer
+    const char* role = SDL_GetHint(SDL_HINT_AUDIO_DEVICE_STREAM_ROLE);
+    if (role == NULL) {
+        PULSEAUDIO_pa_proplist_sets(proplist, PA_PROP_MEDIA_ROLE, "x-maemo");
+    } else {
+        PULSEAUDIO_pa_proplist_sets(proplist, PA_PROP_MEDIA_ROLE, role);
+    }
+#endif
 
     pulseaudio_context = PULSEAUDIO_pa_context_new_with_proplist(mainloop_api, getAppName(), proplist);
     if (!pulseaudio_context) {
